python/lsst/daf/butler/direct_butler/_direct_butler.py

```diff
@@ -1808,22 +1809,58 @@ def ingest(
 
         import_info = self._prepare_ingest_file_datasets(datasets, progress)
 
+        new_datasets, existing_datasets = self._partition_datasets_by_known(datasets)
+        if existing_datasets:
+            if skip_existing:
+                _LOG.info(
+                    "Skipping %d datasets which already exist in the repository.", len(existing_datasets)
+                )
+            else:
+                raise ConflictingDefinitionError(
+                    f"Datastore already contains {len(existing_datasets)} of the given datasets."
+                    f" Example: {existing_datasets[0]}"
+                )
+
         with self.transaction():
             self._ingest_file_datasets(datasets, import_info, progress)
 
             # Bulk-insert everything into Datastore.
             # We do not know if any of the registry entries already existed
-            # (_importDatasets only complains if they exist but differ) so
-            # we have to catch IntegrityError explicitly.
+            # (_importDatasets only complains if they exist but differ).
+            # The _partition_datasets_by_known logic above should catch most
```

Comment:
If concurrent writes are an issue would it help to move this logic inside the transaction?